### ISSUE TITLE
Start creating optimistic transactionThreadReportID and createdReportActionIDForThread and passing it in HoldRequest

### ISSUE THREAD

**Issue by iwiznia** [Expensify/App#52935]
We are about to change some of our backend code to not create the transaction thread reports when we create a new money request. 

In order to do that, we need to ensure that on the action `IOU.putOnHold`, we start creating the data for the transaction thread report and its created action optimistically, the same way we do it for the [requestMoney](https://github.com/Expensify/App/blob/95dc8fcac0de753d1a21608bf02ab487821ca7d0/src/libs/actions/IOU.ts#L3527) action and pass the transactionThreadReportID and createdReportActionIDForThread to the API request.

I also see that the API.write call for putOnHold is not using the `HOLD_MONEY_REQUEST` constant, so let's fix that too.

**Comment by FitseTLT** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Start creating optimistic transactionThreadReportID and createdReportActionIDForThread and passing it in HoldRequest

### What is the root cause of that problem?
New Feature - start creating optimistic transactionThread report and created action for the thread when the transaction thread doesn't exist 
### What changes do you think we should make in order to solve the problem?

First now that we are not guaranteed to always have a transaction thread we will base our navigation on parentReportID and parentReportActionID params
https://github.com/Expensify/App/blob/64eaf2fdd357bd06008adfb9c0c049e09cd375b9/src/ROUTES.ts#L427-L430
```
getRoute: (type: ValueOf<typeof CONST.POLICY.TYPE>, transactionID: string, parentReportID: string, parentReportActionID: string, backTo: string, searchHash?: number) => {
            const route = searchHash
                ? (`${type}/edit/reason/${transactionID}/${searchHash}/?backTo=${backTo}&parentReportID=${parentReportID}&parentReportActionID=${parentReportActionID}` as const)
                : (`${type}/edit/reason/${transactionID}/?backTo=${backTo}&parentReportID=${parentReportID}&parentReportActionID=${parentReportActionID}` as const);
            
```
We should navigate accordingly here
```
 Navigation.navigate(
            // eslint-disable-next-line @typescript-eslint/prefer-nullish-coalescing
            ROUTES.MONEY_REQUEST_HOLD_REASON.getRoute(
                policy?.type ?? CONST.POLICY.TYPE.PERSONAL,
                transactionID,
                moneyRequestReportID ?? '',
                reportAction.reportActionID,
                backTo || activeRoute,
                searchHash,
            ),
```
We will update HoldReasonPage here according to the new params
https://github.com/Expensify/App/blob/64eaf2fdd357bd06008adfb9c0c049e09cd375b9/src/pages/iou/HoldReasonPage.tsx#L39-L47
```
    const {transactionID, parentReportID, parentReportActionID, backTo, searchHash} = route.params;

    const [parentReport] = useOnyx(`${ONYXKEYS.COLLECTION.REPORT}${parentReportID || -1}`);

    // We first check if the report is part of a policy - if not, then it's a personal request (1:1 request)
    // For personal requests, we need to allow both users to put the request on hold
    const isWorkspaceRequest = ReportUtils.isReportInGroupPolicy(parentReport);
    const parentReportAction = ReportActionsUtils.getReportAction(parentReportID ?? '-1', parentReportActionID ?? '-1');

```
and then we will pass only the iou parent action from here
https://github.com/Expensify/App/blob/64eaf2fdd357bd06008adfb9c0c049e09cd375b9/src/pages/iou/HoldReasonPage.tsx#L56
```
        IOU.putOnHold(transactionID, values.comment, parentReportAction, searchHash);

```
Then inside putOnHold we will create the transaction thread and created action if the transaction thread report doesn't exist
```
function putOnHold(transactionID: string, comment: string, iouAction?: OnyxEntry<ReportAction>, searchHash?: number) {
    let reportID = iouAction?.childReportID;
    const optimisticData: OnyxUpdate[] = [];
    if (!reportID) {
        const iouReportID = ReportActionsUtils.getOriginalMessage(iouAction)?.IOUReportID ?? 0;
        const iouReport = ReportUtils.getReport(iouReportID);
        const transactionThread = ReportUtils.buildTransactionThread(iouAction, iouReport);
        reportID = transactionThread.reportID;
        const createdActionForTransactionThread = ReportUtils.buildOptimisticCreatedReportAction(currentUserEmail);
        optimisticData.push(
            {
                onyxMethod: Onyx.METHOD.MERGE,
                key: `${ONYXKEYS.COLLECTION.REPORT}${reportID}`,
                value: transactionThread,
            },
            {
                onyxMethod: Onyx.METHOD.MERGE,
                key: `${ONYXKEYS.COLLECTION.REPORT_ACTIONS}${transactionThread}`,
                value: {
                    [createdActionForTransactionThread.reportActionID]: createdActionForTransactionThread as ReportAction,
                },
            },
        );
    }

```
and also pass the transaction thread report id and created action id as a param here

https://github.com/Expensify/App/blob/64eaf2fdd357bd06008adfb9c0c049e09cd375b9/src/libs/actions/IOU.ts#L7959-L7964
And on more thing is our areHoldRequirementsMet will be false here for our case 
https://github.com/Expensify/App/blob/64eaf2fdd357bd06008adfb9c0c049e09cd375b9/src/pages/home/report/ContextMenu/BaseReportActionContextMenu.tsx#L187-L188
as the calculation depends on the child report of the report action where the context menu is displayed for but in our case there is no transaction thread so we need to determine it only from the report action :
https://github.com/Expensify/App/blob/64eaf2fdd357bd06008adfb9c0c049e09cd375b9/src/pages/home/report/ContextMenu/BaseReportActionContextMenu.tsx#L185
```

    const isParentActionOfRequest = !isEmptyObject(parentReportAction) && ReportActionsUtils.isTransactionThread(parentReportAction);
    const isMoneyRequestOrReport = isMoneyRequestReport || isSingleTransactionView || isParentActionOfRequest;

```
### What alternative solutions did you explore? (Optional)

**Comment by iwiznia** [comment]
@parasharrajat can you review the proposal please?

**Comment by muttmuure** [comment]
@parasharrajat have you had a chance to look at the proposal yet? Could you give an ETA for when you can take a look?

**Comment by iwiznia** [comment]
Bump @parasharrajat

**Comment by muttmuure** [comment]
@parasharrajat hope you're feeling better! Do you need me to reassign this?

**Comment by muttmuure** [comment]
@allroundexperts please can you review https://github.com/Expensify/App/issues/52935#issuecomment-2506838441?

**Comment by allroundexperts** [comment]
Sure. On it now!

**Comment by allroundexperts** [comment]
@iwiznia Why do we need to create the transaction thread on the `IOU.putOnHold` action? Shouldn't we create the transaction thread and action optimistically when a money request is created instead?

It seems to me from looking at your proposal @FitseTLT that we won't be creating the transaction thread and the action till the user puts the request on hold. What would happen in the case when the user clicks on the IOU action in the IOU report? What will we do at that point?

**Comment by FitseTLT** [comment]
>It seems to me from looking at your proposal @FitseTLT that we won't be creating the transaction thread and the action till the user puts the request on hold. What would happen in the case when the user clicks on the IOU action in the IOU report? What will we do at that point?

Obviously, OpenReport API will be called to create the transaction thread report.

**Comment by allroundexperts** [comment]
> Obviously, OpenReport API will be called to create the transaction thread report.

Hm... but since the optimistic transaction thread has not been created, how would the `OpenReport` API create the transaction thread?

**Comment by FitseTLT** [comment]
> > Obviously, OpenReport API will be called to create the transaction thread report.
> 
> Hm... but since the optimistic transaction thread has not been created, how would the `OpenReport` API create the transaction thread?

One of the purpose of Open Report is to create the report if it doesn't exist like in the case of reply in thread
https://github.com/Expensify/App/blob/430110857357fe9e65febe0e653433ffdd713e65/src/libs/actions/Report.ts#L1179-L1199

**Comment by allroundexperts** [comment]
@FitseTLT Are you sure that this will create a transaction thread without any changes?

**Comment by FitseTLT** [comment]
> @FitseTLT Are you sure that this will create a transaction thread without any changes?

I think the fetchReport in report screen here will handle the openReport
https://github.com/Expensify/App/blob/430110857357fe9e65febe0e653433ffdd713e65/src/pages/home/ReportScreen.tsx#L547-L557

**Comment by FitseTLT** [comment]
But let's wait for @iwiznia response on the plan on how to proceed 👍

**Comment by iwiznia** [comment]
> Why do we need to create the transaction thread on the IOU.putOnHold action? Shouldn't we create the transaction thread and action optimistically when a money request is created instead?

No, that's what we are trying to change. Most of the times, the transaction thread report will not be used at all, so we are trying to create it only when it is needed instead of proactively creating it when the money request is created.

**Comment by allroundexperts** [comment]
Reviewing again today!

**Comment by allroundexperts** [comment]
Okay. I went through @FitseTLT's proposal in more detail today again. My only concern was on how we would be creating a transaction thread report if the user clicked on the IOU action. For that case, `navigateToAndOpenChildReport` function should handle creation of the optimistic thread report and action. If it does not, we can modify it to handle that case. I think we can handle that in the PR, and the proposal seems clear enough to me for starting this.

@FitseTLT's [proposal](https://github.com/Expensify/App/issues/52935#issuecomment-2506838441) LGTM!

🎀 👀 🎀 C+ reviewed

**Comment by iwiznia** [comment]
> First now that we are not guaranteed to always have a transaction thread we will base our navigation on parentReportID and parentReportActionID params

Not sure I get why we need this. Can't we generate the URL with the optimistic reportID that will be created when you open it?

**Comment by iwiznia** [comment]
Fixed a bunch of tests, but some still remain

**Comment by iwiznia** [comment]
@FitseTLT bump on the question above

**Comment by FitseTLT** [comment]
> @FitseTLT bump on the question above

I am sorry @iwiznia you should have mentioned me I thought the question was for the reviewer. 
So now, we use the transactionThread report to get the parent report action here
https://github.com/Expensify/App/blob/b134c002463e5726c2792e53de7fe68405f652fa/src/pages/iou/HoldReasonPage.tsx#L28-L33
but if the transaction thread report doesn't exist and user pressed on hold from context menu, my assumption is that the Transaction thread will not be available. Will it be??

**Comment by FitseTLT** [comment]
> > but if the transaction thread report doesn't exist and user pressed on hold from context menu, my assumption is that the Transaction thread will not be available. Will it be??
> 
> Won't be available till the hold request is finished, but I think in that case, we can skip [this logic](https://github.com/Expensify/App/blob/b134c002463e5726c2792e53de7fe68405f652fa/src/pages/iou/HoldReasonPage.tsx#L39) by checking `parentReportAction` is null and go straight to calling `IOU.putOnHold`.

So you are saying the parentAction.childReportID (transactionThreadReportID) will be available even if the transactionThreadREport is not available? @iwiznia

**Comment by iwiznia** [comment]
I don't think that's what I am saying, but I am a bit lost on the question now. Where is `parentAction.childReportID` in that code? 😕

**Comment by FitseTLT** [comment]
That would be here @iwiznia 
https://github.com/Expensify/App/blob/7ea58c74906bf0dcf2b65a10603b9decb7c7fc65/src/libs/ReportUtils.ts#L3461-L3463

how would you navigate to the hold reason page in the first place?

**Comment by iwiznia** [comment]
> That would be here @iwiznia

When/how is that code used in this flow?

> how would you navigate to the hold reason page in the first place?

By doing right click > hold

**Comment by FitseTLT** [comment]
>When/how is that code used in this flow?

When doing right click > hold then `changeMoneyRequestHoldStatus` will be called @iwiznia here

https://github.com/Expensify/App/blob/7ea58c74906bf0dcf2b65a10603b9decb7c7fc65/src/pages/home/report/ContextMenu/ContextMenuActions.tsx#L296

**Comment by iwiznia** [comment]
ok, so since we will be generating an optimistic reportID I think that should work, no?

**Comment by FitseTLT** [comment]
> ok, so since we will be generating an optimistic reportID I think that should work, no?

If we skip [this logic](https://github.com/Expensify/App/blob/b134c002463e5726c2792e53de7fe68405f652fa/src/pages/iou/HoldReasonPage.tsx#L39) and [this](https://github.com/FitseTLT/App/blob/cbf6112f3fb944e5bf02cdf4062a60f70f452929/src/pages/iou/HoldReasonPage.tsx#L57) yeah it can work.

**Comment by iwiznia** [comment]
Cool, let's do that then

**Comment by iwiznia** [comment]
Any update @FitseTLT?

**Comment by muttmuure** [comment]
bump @FitseTLT

**Comment by iwiznia** [comment]
@muttmuure let's reassign this issue to someone else?

**Comment by FitseTLT** [comment]
Sorry guys for the delay. I will get back to office next week and make pr ready hopefully by Monday.

**Comment by iwiznia** [comment]
@FitseTLT are you going to manage to send the PR today?

**Comment by FitseTLT** [comment]
@iwiznia I was working on this but encountered a problem.

For the case where there is no transaction thread attached to a money request created yet we are planning  on having childReportID populated for the parent report action, right? But this causes one problem: we can't identify the transaction thread report doesn't exist because it doesn't totally exist or it exists but not available in onyx (because the user hasn't opened the report yet after signing in).

**Comment by iwiznia** [comment]
Not sure I follow what the problem is... I think it's the same as when you create a chat with 2 users and set an optimisticID. The server would create a report with the passed reportID if the parent does not have a child or ignore the optimisticID and not create any report if the parent did have a child.

**Comment by FitseTLT** [comment]
> Not sure I follow what the problem is... I think it's the same as when you create a chat with 2 users and set an optimisticID. The server would create a report with the passed reportID if the parent does not have a child or ignore the optimisticID and not create any report if the parent did have a child.

Okay but the BE also need to reset createdReportAction for the transaction thread for the case where there is no need to create a transaction thread because there is no way to know that from FE. I have created a draft [PR](https://github.com/Expensify/App/pull/55491) so do let me know when you are done with the BE changes so that I can test and finish up 👍

**Comment by iwiznia** [comment]
Hmmmm yeah, isn't there something already in our code to handle optimistic reportIDs that do not match the ones coming from the backend?

**Comment by muttmuure** [comment]
@FitseTLT do you have an ETA for your PR to be ready for review?

**Comment by FitseTLT** [comment]
@muttmuure Draft pr is [here](https://github.com/Expensify/App/pull/55491) waiting for BE changes 👍

**Comment by muttmuure** [comment]
@iwiznia is this waiting on you then?

**Comment by iwiznia** [comment]
Maybe, but @FitseTLT can you answer this https://github.com/Expensify/App/issues/52935#issuecomment-2603321534?

**Comment by FitseTLT** [comment]
> Hmmmm yeah, isn't there something already in our code to handle optimistic reportIDs that do not match the ones coming from the backend?

@iwiznia I tried to check how we are handling this for chat reports now. We first open an optimistic report with a random report id and OpenReport API will populate onyx with
```
{
    "onyxData": [
        {
            "onyxMethod": "merge",
            "key": "report_8089554103859171",
            "value": {
                "reportID": "8089554103859171",
                "preexistingReportID": "8191166395709480"
            }
        }
    ],
    "jsonCode": 200,
    "requestID": "90e5b4f42d9c7c52-SJC"
}
```

to set the report with `preexistingReportID` and we handle unused optimistic id here
https://github.com/Expensify/App/blob/f1c826d5c993ef8beab0ac7597b8be2f6eb277cd/src/libs/Middleware/HandleUnusedOptimisticID.ts#L9-L25

**Comment by iwiznia** [comment]
@FitseTLT bump https://github.com/Expensify/App/issues/52935#issuecomment-2657726700

**Comment by FitseTLT** [comment]
> Ah ok, so if we send back `preexistingReportID` in the value of a merge onyx method for the report key, it will take care of everything? In what api exactly would we need to add it?

@iwiznia This approach doesn't work for our cases. Here is how we will handle this:

On the money request preview parent report action we should have a way to identify if a transaction thread linked to it exists or not. We can do that by not populating it with `childReportID`.
So if childReportID doesn't exist we know that there is no transaction thread linked to it. Therefore, on hold we will create the transaction thread optimistically and pass the report id and created report action id as a param. BTW we should do the same when the user presses the money request preview. We can identify that there is no transaction thread if `childReportID` is undefined and optimistically create the transaction thread and pass the params to may be a new API or re-use `OpenReport` API.

**Comment by iwiznia** [comment]
> On the money request preview parent report action we should have a way to identify if a transaction thread linked to it exists or not. We can do that by not populating it with childReportID.

I think this should happen automatically already once we deploy the PR to not create the threads proactively

**Comment by iwiznia** [comment]
> I think this should happen automatically already once we deploy the PR to not create the threads proactively.

This will get deployed today/tomorrow. @allroundexperts you will need to stop sending the optimistic transactionThreadID in the RequestMoney api request in order for the report to not be created proactively. LMK if it works or you need some help.

**Comment by JmillsExpensify** [comment]
Thanks for the update. Not sure if I'm the one that's supposed to comment to remove the overdue?

**Comment by iwiznia** [comment]
Oh wait, we had to revert the PR I mentioned because it caused issues and I haven't sent the new PR again.

**Comment by iwiznia** [comment]
The new PR is up for review

**Comment by JmillsExpensify** [comment]
PR up for review

**Comment by mallenexpensify** [comment]
Are any of these held on another? 
- https://github.com/Expensify/App/issues/52935
- https://github.com/Expensify/App/issues/58828
- https://github.com/Expensify/Expensify/issues/433103

For this issue, what's the PR we're waiting to merge?  Thx

**Comment by iwiznia** [comment]
We are waiting for this to get deployed https://github.com/Expensify/Web-Expensify/pull/45944 
Which should be done today/tomorrow to staging, which is good enough for @FitseTLT to start testing

**Comment by JmillsExpensify** [comment]
@FitseTLT have you been able to test yet?

**Comment by FitseTLT** [comment]
@iwiznia Is it ready for test? If yes how should I test? And also, I was wondering if what we are doing here has any relation with beta TableReportView?

**Comment by iwiznia** [comment]
> @iwiznia Is it ready for test? If

yes, it should be

> If yes how should I test? 

Make sure you are in the latest code (which should not create an optimistic thread when calling RequestMoney) and then put the transaction on hold.

**Comment by iwiznia** [comment]
>  And also, I was wondering if what we are doing here has any relation with beta TableReportView?

I think not?

**Comment by allroundexperts** [comment]
@FitseTLT Do you have an update here?

**Comment by JmillsExpensify** [comment]
Good question. I agree that it'd be good to pick this back up.
